src/inventory.rs

```diff
@@ -0,0 +1,303 @@
+use std::collections::HashSet;
+use std::ops::Range;
+
+use rayon::prelude::ParallelIterator;
+
+use crate::client::Clients;
+use crate::config::Config;
+use crate::protocol::packets::s2c::play::SetContainerContent;
+use crate::protocol::{Slot, SlotId, VarInt};
+use crate::slab_versioned::{Key, VersionedSlab};
+
+pub trait Inventory {
+    fn slot(&self, slot_id: SlotId) -> Slot;
+    fn set_slot(&mut self, slot_id: SlotId, slot: Slot);
+    fn slot_count(&self) -> usize;
+    fn mark_dirty(&mut self, dirty: bool);
+    fn is_dirty(&self) -> bool;
+
+    // TODO: `entry()` style api
+
+    fn slots(&self) -> Vec<Slot> {
+        (0..self.slot_count())
+            .map(|s| self.slot(s as SlotId))
+            .collect()
+    }
+
+    fn consume_one(&mut self, slot_id: SlotId) {
```

Comment:
Probably shouldn't have a dedicated function in the trait to do this.

Reply:
I agree, but I'm intentionally not focusing on the user API right now. I fully intend to change this later.